src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs

```diff
@@ -588,10 +1402,24 @@ internal static bool Dispatch()
                         System.Diagnostics.Tracing.FrameworkEventSource.Log.ThreadPoolDequeueWorkObject(workItem);
 
                     //
-                    // If we found work, there may be more work.  Ask for another thread so that the other work can be processed
-                    // in parallel.  Note that this will only ask for a max of #procs threads, so it's safe to call it for every dequeue.
+                    // We are about to execute external code, which can take a while, block or even wait on something from other tasks.
+                    // Make sure there is a request, in case we do not come back for long while.
                     //
-                    workQueue.EnsureThreadRequested();
+                    // If we execute more than one task per quantum, ask for an aditional thread (up to #procs) on the first success and 
+                    // every 1024 successes thereafter.
+                    // NB: we are not very sensitive to the number. It is here just to throttle the rate of additional expansion requests. 
+                    //     The number of cycles per quantum can be in 100K range if tasks are "easy", so no need to be too aggressive here. 
+                    //     New workers will ask for more spinners if there is work, and stay around themselves, 
+                    //     so #procs is saturated quickly when pool is busy.                    
+                    if ((dequeueCycles++ & 1023) == 0)
+                    {
+                        maxdeq = dequeueCycles;
+                        workQueue.RequestThread();
```

Comment:
Similar to what I said in my comment in `EnsureThreadRequested()`, this also has the effect of serializing the release of waiting thread pool threads. Suppose an app requested a minimum of 100 threads to avoid starvation or to prefer latency over throughput. Suppose that the app gets a burst of 10,000 work items all at once. It would take a relatively long time for all 100 waiting threads to be released. It would actually be worse than single-threaded serialization because beyond releasing proc count threads, each additional waiting thread that is woken has to also wait for a time slice in order to be scheduled and to release one more thread. As a result the delay would also scale worse with the number of worker threads.

I mentioned in the other comment as well, but for now I would suggest to revert the thread requesting logic to how it was before.

Reply:
yes, lets change back

---
In reply to: [250424672](https://github.com/dotnet/coreclr/pull/18403#discussion_r250424672) [](ancestors = 250424672)